tmt/steps/discover/__init__.py

```diff
@@ -131,6 +132,37 @@ def log_import_plan_details(self) -> None:
             for key, value in cast(dict[str, str], remote_plan_id.to_minimal_spec()).items():
                 self.verbose(f'import {key}', value, 'green')
 
+    def filter_for_rerun(self) -> None:
+        """ Filter out passed tests from previous run data """
+        assert isinstance(self.step.parent, tmt.base.Plan)  # narrow type
+        old_results: Path = self.step.parent.last_run_execute / 'results.yaml'
+        results = [
+            Result.from_serialized(data) for data in
+            tmt.utils.yaml_to_list(self.read(old_results))]
+        results_failed: list[str] = []
+        results_passed: list[Result] = []
+        for result in results:
+            if (
+                    result.result is not tmt.result.ResultOutcome.PASS and
+                    result.result is not tmt.result.ResultOutcome.INFO):
+                results_failed.append(result.name)
+            else:
+                results_passed.append(result)
+
+        # Overwrite previous run results to only include passed cases
+        self.debug(
+            f"Overwriting {old_results} to only include passed results: "
+            f"{', '.join([result.name for result in results_passed])}")
+        self.write(
+            old_results,
+            tmt.utils.dict_to_yaml([result.to_serialized() for result in results_passed]))
+
+        tests_to_execute: list[tmt.base.Test] = []
+        for test in self._tests:
+            if test.name in results_failed:
+                tests_to_execute.append(test)
+        self._tests: list[tmt.base.Test] = tests_to_execute
```

Comment:
I'm not sure the name test would be enough. We're in the context of a single `discover` phase, so all tests share `where`, that's fine,  but one test can still be repeated multiple times if I'm not mistaken:

```yaml
how: shell
tests:
  - name: /foo
    test: tmt --help
  - name: /foo
    test: tmt --help
```

Probably of not much use, but possible? I think the filter needs to consider also a serial number besides the test name. If the results for the same test *name* and *serial number* failed, then the test instance would pass through.

Reply:
can I rely on the serial number to be the same in between runs? I guess the order should be always the same and the filtering is purposefully done after the serial number generation so I guess yeah?

I wanted to have it more readable and was expecting more conditions and checks, but looking at your example it's same level of readability so...thanks, I'll use that :)